[add] cms_status_message

simahawk opened the pull request:
Splitted "status message" feature out of website_cms to make it reusable without the whole package.

@Yajo 1st: wish you an happy and productive new year :) 2nd: your feedback here is much appreciated :)

simahawk commented:
@Yajo @antespi @lasley @PCatinean your feedback is much appreciated here :)

yajo commented on cms_status_message/__openerp__.py:
```diff
@@ -0,0 +1,18 @@
+{
+    'name': 'CMS status message',
+    'summary': """
+        Basic status messages for your CMS system""",
```
Yikes, ugly string here 🙃

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,36 @@
+        'info': _('Info'),
+        'success': _('Success'),
+        'danger': _('Error'),
+        'warning': _('Warning'),
```
This is wrong. When you do translations like this, they get translated at class declaration time, AKA in English always.

You have to declare a method, which will be called at request time, with current user's lang in its context.

BTW, I'd add an `assert type_ in types` here when you add the method.

simahawk replied:
I don't want to force users to stick to those types. Potentially you could pass `type_='foo'` to have a custom alert box with class `alert-foo`.

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @api.model
+    def add_status_message(self, msg, mtitle='', mtype='info'):
```
Maybe you don't want an `mtitle` in your message, so I'd put here `mtitle=None`, and below:

```
mtitle = self.default_status_msg_title(mtype) if mtitle is None else mtitle
```

...  to allow the developer to pass `mtitle=""` and remove the title part. A docstring on this would help.

Besides... why the `m` in `mtitle` and `mtype`? Isn't it obvious enough that it is for the message? 😉 ([`mtype` should be `type_`](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles), BTW).

simahawk replied:
fixed.

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,36 @@
+    def get_status_message(self):
+        if request.session:
+            return request.session.pop('status_message', {})
+        return {}
```
This implementation is wrong. You are always returning the dict of the last message, although in the view you expect to get a list of messages. Should be:

    def pop_status_messages(self):
        try:
            return request.session.pop('status_message', [])
        except AttributeError:
            return []

simahawk replied:
fixed.

yajo commented on cms_status_message/README.rst:
```diff
@@ -0,0 +1,97 @@
+            <t t-call="cms_status_message.status_message" />
+        </xpath>
+
+    </template>
```
This should be included by default. You can set an optional template and let users disable it, but add it by default please.

simahawk replied:
was my 1st idea then I decided to not decide for everyone... but it makes sense by default ;P DONE!

yajo commented on cms_status_message/README.rst:
```diff
@@ -0,0 +1,97 @@
+    * 'info': 'Info'
+    * 'success': 'Success'
+    * 'danger': 'Error'
+    * 'warning': 'Warning'
```
Remove indentation here.

simahawk replied:
fixed.

yajo commented on cms_status_message/templates.xml:
```diff
@@ -0,0 +1,21 @@
+                  <span t-translation="off" aria-hidden="true" class="fa fa-remove"></span>
+                </button>
+                <strong t-esc="msg['title']" />
+                <t t-esc="msg['msg']" />
```
Put this in a `<span>` please, it will ease the task of styling it under some scenarios.

simahawk replied:
addes classes too.

yajo commented on cms_status_message/templates.xml:
```diff
@@ -0,0 +1,21 @@
+                <button type="button" class="close" data-dismiss="alert" aria-label="Close">
+                  <span t-translation="off" aria-hidden="true" class="fa fa-remove"></span>
+                </button>
+                <strong t-esc="msg['title']" />
```
Add `t-if="msg['title']"`

simahawk replied:
done.

yajo reviewed (changes requested):
Awesome!

Not sure about the addon name... `website_status_message` would be good, and the website repo too. This is not cms-specific IMHO. Not a requirement though.

OTOH please add some tests. [Look here for an idea.](https://github.com/OCA/website/pull/298/files#diff-fb24fdad3fe870ba283a7d07581a91e6)

simahawk commented:
@Yajo thanks for the pointers, especially for the tests: I was looking precisely for something like that ;)
WIP!

simahawk committed:
[add] cms_status_message

lasley reviewed (commented):
Looks good @simahawk! @Yajo's note about the ugly string in manifest still looks unattended to though.

Any thoughts on tests? There's a lot of red lines here that make me 😢  😉

simahawk commented:
@lasley still a WIP for some fixes + tests ;) but I cannot add a WIP label :(

yajo commented:
Labeled. Please tell when WIP ends 😉

simahawk commented:
@lasley  added some tests. I'd like to test also rendering but the HttpCase does not provide a request (see this comment too https://github.com/odoo/odoo/blob/9.0/openerp/tests/common.py#L287) which is really a pity (and crapy, I'd say). So, is impossible to render the template, even if you render it directly from template obj.

Sounds like you are forced to mock it. The only place where I found a proper mock for such stuff is your work :) Here: https://github.com/OCA/server-tools/blob/9.0/password_security/tests/test_password_security_home.py#L44.

But WTH: I only need a website+session attached to the request! Any tip?

yajo commented:
If you want to test the rendering, then use a tour.

simahawk commented:
@lasley @Yajo in the end I found a simple mock approach (thanks Dave for that piece of code! :)).
So:
* added tests, for rendering too
* added new option "dismissible" to control alert dismission

@Yajo you can remove WIP label if tests pass.

yajo commented on cms_status_message/__openerp__.py:
```diff
@@ -0,0 +1,17 @@
+        'website',
+    ],
+    'data': [
+        'templates.xml',
```
This must go in the `templates/` folder.

simahawk replied:
ok, I thought it was fine w/ just one file ;)

simahawk replied:
according to this "models.py" is not good neither... isn't it?

yajo replied:
Indeed.

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,65 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import models, api
+from openerp.http import request
+from openerp import _
```
Why not `from openerp import _, api, models` above?

simahawk replied:
usually I don't do more than one import on one line, because it's easier to copy/paste what you need and above all to spot import issues. But I can merge them if needed.

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,65 @@
+class Website(models.Model):
+    _inherit = 'website'
+
+    @property
```
I'm not sure this will work fine... Did you check if strings get correctly translated? I also don't know if this can be easily inherited/extended... 🤔

simahawk replied:
In my experience inheritance works... about translations: I'll add Italian translation and test it. Sounds good? :)

simahawk replied:
added tests for translations + test: it works ;)
Only one doubt: if you customize translations in your own module this test is going to fail?
In my experience this is true for instance with tour tests: if you remove the "add content" menu in your theme, the tour that test page add is going to fail and there's no way to deactivate it than patch the phantom test (or remove the whole JS from tests assets probably).

yajo replied:
You can test if the language is installed and skip it otherwise.

yajo replied:
You have to fix Travis. I think you can remove the translation test, I was meaning to do a manual check. Translations is a builtin framework system that should be tested in the framework itself. I was just really asking if it worked under `@property` methods, but given it does, it's enough to know it IMHO 😉

simahawk replied:
oooook :) Green now!

yajo commented on cms_status_message/models.py:
```diff
@@ -0,0 +1,65 @@
+                           dismissible=True, session=None):
+        """Inject status message in session.
+
+        @param msg: the message you want to display.
```
The standard for docstrings is [to use the sphinx syntax](http://www.sphinx-doc.org/en/1.5.1/domains.html#info-field-lists) AFAIK. Not requirement though.

simahawk replied:
updated

yajo reviewed (changes requested):
One little requirement and some questions...

yajo reviewed (changes requested):
Just apply [common module structure](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#directories) and you'll finally get my 👍 😉

simahawk commented:
@Yajo done and build is green ;)

simahawk commented:
@Yajo @lasley someone forgot to push the "merge" button? :laughing:
